tests/PHPStan/Analyser/nsrt/array_splice.php

```diff
@@ -21,41 +21,261 @@ function insertViaArraySplice(array $arr): void
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	array_splice($arr, 0, null, ['hello']);
+	assertType('array{\'hello\'}', $arr);
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	array_splice($arr, 0);
+	assertType('array{}', $arr);
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	/** @var array<\stdClass> $arr2 */
+	array_splice($arr, 1, 1, $arr2);
+	assertType('non-empty-array<int<0, max>, \'baz\'|\'foo\'|stdClass>', $arr);
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	/** @var array<\stdClass> $arr2 */
+	array_splice($arr, 0, 1, $arr2);
+	assertType('non-empty-array<\'b\'|int<0, max>, \'bar\'|\'baz\'|stdClass>', $arr);
+}
+
+function constantArraysWithOptionalKeys(array $arr): void
+{
+	/**
+	 * @see https://3v4l.org/jrqoZ
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b?: 1, c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/lbUJG
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a: 0, 0: \'hello\', c: 2}|array{b: 1, 0: \'hello\'}', $arr);
+
+	/**
+	 * @see https://3v4l.org/7uPmV
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a?: 0, b: 1, 0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/hB8pG
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/TjfHT
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b?: 1, c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/D8PSE
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a: 0, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/8RfDs
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a: 0, b?: 1, 0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/sPfpN
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/Ddpku
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b: 1, c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/O4LLi
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a: 0, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/QQO84
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a: 0, b: 1, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/K5RDp
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', b: 1}|array{0: \'hello\', c: 2}', $arr);
```

Comment:
not sure how useful it is that PHPStan fully understands this, but I think it's pretty neat 😊

Reply:
it still understands this, but the return array is not a union anymore because the solution changed